games/static/games/proto.js

```diff
@@ -307,87 +347,146 @@ class Board {
+            if (this.theTopIs(rocketLocation)) {
+                clearInterval(rocketId)
+                this.waitAndEraseRocketAt(rocketLocation) 
             }
-            catch(err) {
-                console.log(rocketInFlight)
+            if (this.squareAt(nextRocketLocation)) {
+                if (this.squareAt(nextRocketLocation).hasAlien()) {            
+                    clearInterval(rocketId)
+                    this.waitAndEraseRocketAt(rocketLocation)
+                    this.reduceAlienHealthByOneAt(nextRocketLocation)
+                    this.updateAlienHitAt(nextRocketLocation)
+                }
             }
+            if (this.squareAt(rocketLocation).hasAlien()) this.squareAt(rocketLocation).setEmpty()
+            this.squareAt(rocketLocation).setCharacter(rocketInFlight)
+
+        }, 100);
+    }
+    updateAlienHitAt(nextRocketLocation) {
+        let alienHit = this.squareAt(nextRocketLocation).pickCharacterUp()
+        this.squareAt(nextRocketLocation).setCharacter(alienHit)
+    }
+    reduceAlienHealthByOneAt(nextRocketLocation) {
+        this.squareAt(nextRocketLocation).character.health -= 1
+    }
+    waitAndEraseRocketAt(rocketLocation) {
+        setTimeout(() => { this.liftRocketAt(rocketLocation) }, 100)
+    }
+    theTopIs(rocketLocation) {
+        return this.squareAt(rocketLocation).isTop()
+    }
+    update(nextRocketLocation) {
+        return nextRocketLocation + this.directionUp()
+    }
+    updateIndexOf(rocket) {
+        rocket.index = rocket.index + this.directionUp()
+    }
+    liftRocketAt(rocketLocation) {
+        return this.squareAt(rocketLocation).pickCharacterUp()
+    }
+    getNextIndexOf(rocket) {
+        return rocket.index + this.directionUp()
+    }
+    getIndexOf(rocket) {
+        return rocket.index
+    }
+    createRocket() {
+        return new Rocket('rocket', this.squareAboveDefender(), this)
+    }
+    
+    squareAboveDefender() {
+        return this.defenderLocation - this.width
+    }
+
+    squareAt(rocketLocation) {
+        return this.squares[rocketLocation]
+    }
+
+    directionUp() {
+        return -this.width
+    }
+    
+}
+function calcNextAlienMove(previousDirection, counter, numberOfMoves, down) {
+    if (counter < numberOfMoves) {
+        return {
+            'counter': increaseByOne(counter),
+            'previousDirection':  previousDirection, 
+            'direction': previousDirection,
+            
         }
-        switch(e.keyCode ) {
-            case 32:
-                rocketId = setInterval(moveRocket , 100 , rocket)
-                break
+    }
+    if (counter === numberOfMoves) {
+        return {
+            'counter': zero(counter),
+            'previousDirection': getTheOppositeOf(previousDirection),
+            'direction': down,
+            
         }
-        
     }
+}
 
-    
+function increaseByOne(counter){
+    return counter += 1
```

Comment:
same here
dont change the value you got
you can return a new calculated value to the caller and let the caller do the mutation